### ISSUE TITLE
::= does not work for rvalue covers because a pointer is expected.

### ISSUE THREAD

**Issue by davidpiuva** [cogneco/ooc-kean#544]
When a FloatBox2D is a value type given from a collection look-up or function call, rightBottom can't be accessed without storing the box to make it a referrable cover because the ::= operand generates a getter that expects a pointer.

A workaround is to store rvalues to make them into cloned lvalues.

This can be solved by getting lvalues from [] operands since rvalues are slow to pass around.
At least, the error should be explained by the ooc compiler instead of failing in the C compiler.

**Comment by davidhesselbom** [comment]
Can you post a non-working code example?

**Comment by davidpiuva** [comment]
// Not working
result = myCovers[i] myAttribute

// Working
myCover := myCovers[i] // Convert rvalue to lvalue
result = myCover myAttribute

**Comment by davidhesselbom** [comment]
Ah, I see what you mean. So collections are not suitable as get properties, then.

**Comment by davidhesselbom** [comment]
Would it help to wrap parentheses around `myCovers[i]`?

**Comment by davidpiuva** [comment]
Parentheses did not help.

**Comment by davidhesselbom** [comment]
How about `result := (myCovers[i] as MyCover) attribute`?

**Comment by davidpiuva** [comment]
Failed too.

**Comment by davidhesselbom** [comment]
Interesting...

**Comment by marcusnaslund** [comment]
I doubt we're going to change rock regarding this. Could you post an issue about it there, though? https://github.com/fasterthanlime/rock/issues
Should we change these properties (potentially quite a lot)? Otherwise, this is not really a bug in kean.

**Comment by davidpiuva** [comment]
It is probably too wide for ooc since .NET has the same dangerous bug/limitation.

**Comment by marcusnaslund** [comment]
You're right, C# doesn't do this either.
But I agree with 

> At least, the error should be explained by the ooc compiler instead of failing in the C compiler.

so that would warrant an issue in rock. Nothing to do here though, it seems.

**Comment by sebastianbaginski** [comment]
Maybe this is a good place to ask. Can you guys tell me what's so good about properties anyway ? Coming from the C++ background, the whole *property* idea seems strange to me and I still can't really get used to it. So basically you have something that looks like a variable ...
```ooc
Example: class {
    property ::= this something + 29
}

var ::= object property
```
... but is in fact a function :)
In C++, you have a const methods instead
```C++
class Example {
public:
    int property() const { return this->something + 29; }
};
auto var = object.property();
```
... and it is clear from the syntax that you are going to call a function.
So the properties in ooc can lead to confusing and ineffective code (this example is real, I fixed it some time ago in vector class)
```ooc
for (i in 0 .. this size)
    this[i] /= this max
```
where `this max` (or something) was a *property* which returned maximum element from the vector.
It is not clear from the syntax that you are going to call a function, in C++ it would be clear that you are doing things in semi-effective way:
```C++
for (int i=0 ; i<this->size() ; ++i) 
    (*this)[i] /= this->max();
```
So, basically that's why I'm having a hard time to switch and keep writing things as methods - because properties are methods under the hood anyway...

**Comment by davidhesselbom** [comment]
@sebastianbaginski It's probably because of our C# background. Maybe this would be a good read for you: https://msdn.microsoft.com/en-us/library/vstudio/ms229054%28v=vs.100%29.aspx

**Comment by sebastianbaginski** [comment]
Thanks for the link, *"Properties are meant to be used like fields"*, this clears some of my doubts.

**Comment by davidhesselbom** [comment]
Also,

> less experienced developers find properties easier to use

I guess that's why you find them so confusing - you have too much experience ;)

**Comment by sebastianbaginski** [comment]
haha ^_^

... ok in order to write something constructive - do we still want to keep this issue open ?

**Comment by davidhesselbom** [comment]
Well, it should have been an issue in rock, because there's nothing we can do about it in ooc-kean other than avoid doing it (and issues are not meant to be left open forever to serve as warnings). Let's leave this open until someone does open an issue in rock.

**Comment by marcusnaslund** [comment]
Done. https://github.com/fasterthanlime/rock/issues/940
